### PR TITLE
Assign tasks to workers

### LINKED ISSUE #411: Keep GPU workload to one worker locally.
**Is your feature request related to a problem? Please describe.**
I have a very small GPU (6Gb), so even with a batch size of 1, I can't run Azimuth with saliency on my computer. Both Dask workers load the model on the GPU which creates a OOM.

**Describe the solution you'd like**
When using a local Dask cluster (the default), it would be great if we assigned resources to one of them so that it handles all GPU-related tasks.

**Describe alternatives you've considered**
I could run on CPU, but that takes a while.

**Additional context**

### PR DESCRIPTION
**Pull request by gabegma**
Part of #521
Fixes #411 

## Description:
* The tests used to be flacky on the CI, but I after rebasing on all the changes, I'm not sure if that is still the case. It seems it's no longer the case given the amount of back and forth. Yay!!
* This allows the user to manually assign the GPU to worker 0, because this is where the pipeline and encoder will be loaded.

Here is the memory before and after the change.
6  before
7  assign workers

## Checklist:

You should check all boxes before the PR is ready. If a box does not apply, check it to acknowledge it.

* [x] **ISSUE NUMBER.** You linked the issue number (Ex: Resolve #XXX).
* [x] **PRE-COMMIT.** You ran pre-commit on all commits, or else, you
  ran `pre-commit run --all-files` at the end.
* [x] **USER CHANGES.** The changes are added to CHANGELOG.md and the documentation, if they impact
  our users.
* [x] **DEV CHANGES.**
    * Update the documentation if this PR changes how to develop/launch on the app.
    * Update the `README` files and our wiki for any big design decisions, if relevant.
    * Add unit tests, docstrings, typing and comments for complex sections.

### CODE REVIEW

File: azimuth/modules/base_classes/dask_module.py [hunk 1]
```diff
@@ -94,6 +94,10 @@ def _get_config_scope(self, config) -> ConfigScope:
             scoped_config = base.__args__[0]
         return cast(ConfigScope, scoped_config.parse_obj(config.dict(by_alias=True)))
 
+    @property
+    def can_load_model(self) -> bool:
+        raise NotImplementedError
```
Comment (gabegma) [inline azimuth/modules/base_classes/dask_module.py:99]: I had import circular issues if I implemented the property here.

File: azimuth/modules/base_classes/mixin.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class ModelMixin:
+    """Mixin to add to modules that can load the model.
+
+    This Mixin exists to assign tasks that can load the model to the same worker, to avoid
+    loading the model by 2 workers simultaneously.
+    """
+
+    config: ConfigScope  # type: ignore
```
Comment (gabegma) [inline azimuth/modules/base_classes/mixin.py:35]: I couldn't figure out how to solve for the `mypy` issue.

Comment (Dref360) [reply azimuth/modules/base_classes/mixin.py:35]: What was the error?

Comment (gabegma) [reply azimuth/modules/base_classes/mixin.py:35]: `Type variable "azimuth.modules.base_classes.dask_module.ConfigScope" is unbound`

File: azimuth/modules/base_classes/dask_module.py (hunk 1 above)
Comment (Dref360) [reply azimuth/modules/base_classes/dask_module.py:99]: Right, we define ConfigScope in `dask_module.py`, but it is used in the Mixin. Maybe move it?

Comment (gabegma) [reply azimuth/modules/base_classes/dask_module.py:99]: Ok, I needed to move another thing around but it now works.

File: azimuth/modules/base_classes/mixin.py (hunk 2 above)
Comment (Dref360) [reply azimuth/modules/base_classes/mixin.py:35]: At this moment, it "must" be a ModelContractConfig right?

Comment (gabegma) [reply azimuth/modules/base_classes/mixin.py]: As discussed offline, I had a `mypy` error with ModelContractConfig too. I'll revert the Mixin and use a class variable instead.

File: azimuth/modules/base_classes/module.py [hunk 3]
```diff
@@ -164,6 +164,8 @@ def get_model(self):
         Raises:
             ValueError if no valid pipeline exists.
         """
+        if not self.can_load_model:
+            raise ValueError("This module cannot load the model. Set can_load_model to True.")
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/module.py:168]: Very good idea to catch it here! 👍 
Detail, but to be rigorous, it seems a [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError) is not quite appropriate here:
> Raised when an operation or function receives an argument that has the right type but an inappropriate value

I think we can use a more generic [`RuntimeError`](https://docs.python.org/3/library/exceptions.html#RuntimeError):
> Raised when an error is detected that doesn’t fall in any of the other categories.

```suggestion
            raise RuntimeError("This module cannot load the model. Set can_load_model to True.")
```

File: azimuth/modules/base_classes/dask_module.py [hunk 4]
```diff
@@ -37,6 +37,7 @@ class DaskModule(HDF5CacheMixin, Generic[ConfigScope]):
     """
 
     allowed_splits = {DatasetSplitName.train, DatasetSplitName.eval}
+    can_load_model: bool = False
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/dask_module.py:40]: Regarding the "can" in the attribute name, is it because there are some modules that might or might not actually load the model? If so, I think it would help to leave a comment on why it's `can_load_model` and not something like `loads_model`, `needs_model`, or `needs_to_load_model`.
Asked differently, is there a reason to set `can_load_model = True` and not load the model?

File: azimuth/modules/base_classes/dask_module.py [hunk 5]
```diff
@@ -118,6 +119,7 @@ def start_task_on_dataset_split(
             pure=False,
             dependencies=deps,
             key=f"{self.task_id}_{uuid.uuid4()}",  # Unique identifier
+            workers=0 if self.can_load_model else None,
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/dask_module.py]: > This allows the user to manually assign the GPU to worker 0

How does one do that? Should we do something in our code?

File: azimuth/modules/dataset_analysis/similarity_analysis.py [hunk 6]
```diff
@@ -45,10 +44,7 @@ def get_encoder_name_or_path(self):
         return model_name_or_path
 
     def get_encoder(self):
```
Comment (JosephMarinier) [inline azimuth/modules/dataset_analysis/similarity_analysis.py:46]: Should we raise an exception if `not self.can_load_encoder`? I understand it is not possible now since `get_encoder()` is only defined in `FAISSModule`, which has `can_load_encoder = True`, but it would be an extra safety net for if someone tries to create, for some reason, a new module that inherits `FAISSModule` and overwrites `can_load_encoder = False`.

File: azimuth/modules/base_classes/artifact_manager.py [hunk 7]
```diff
@@ -158,6 +160,11 @@ def get_metric(self, config, name: str, **kwargs):
             self.metrics[metric_hash] = load_custom_object(config.metrics[name], **kwargs)
         return self.metrics[metric_hash]
 
+    def get_encoder(self, encoder_name_or_path):
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/artifact_manager.py:163]: Why
> Move the encoder to ArtifactManager

No issue, I just want to follow. 😅

File: azimuth/modules/base_classes/dask_module.py [hunk 8]
```diff
@@ -119,7 +120,7 @@ def start_task_on_dataset_split(
             pure=False,
             dependencies=deps,
             key=f"{self.task_id}_{uuid.uuid4()}",  # Unique identifier
-            workers=0 if self.can_load_model else None,
+            workers=0 if self.can_load_model else 1 if self.can_load_encoder else None,
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/dask_module.py:123]: So, we don't want a module to have both `can_load_model = True` and `can_load_encoder = True`, right?
I have an idea to have typing enforce that. 💡 I haven't tested, though.
We create a `Worker` enum to give workers meaningful names:
```py
class Worker(IntEnum):
    model = 0
    encoder = 1
```
We let modules set on which worker they should run:
```diff
 class DaskModule(HDF5CacheMixin, Generic[ConfigScope]):
 ...
-    can_load_model: bool = False
-    can_load_encoder: bool = False
+    worker: Optional[Worker] = None
```
```diff
 class FAISSModule(IndexableModule[SimilarityConfig]):
 ...
-    can_load_encoder = True
+    worker = Worker.encoder
```
We are still able to raise an exception if it's not set properly:
```diff
     def get_model(self):
     ...
-        if not self.can_load_model:
+        if self.worker != Worker.model:
            raise RuntimeError("This module cannot load the model. Set can_load_model to True.")
```
And here, we simply pass the worker:
```suggestion
            workers=self.worker
```
What do you think?

File: azimuth/modules/base_classes/module.py (hunk 3 above)
Comment (gabegma) [reply azimuth/modules/base_classes/module.py:168]: Perfect, I didn’t know what to put. 

It’s worth saying that this method (`can_load_model`) is not perfect, a module can call the prediction module with `.compute()` or `.compute_on_dataset_split()`, which will load the model without calling `get_model()`, the best example of this is perturbation testing. I’m not sure how we can make that more fail-proof. Perhaps I can add a TODO.

File: azimuth/modules/base_classes/dask_module.py (hunk 4 above)
Comment (gabegma) [reply azimuth/modules/base_classes/dask_module.py:40]: Right, we could rename to `loads_model`, good point.

File: azimuth/modules/base_classes/dask_module.py (hunk 5 above)
Comment (gabegma) [reply azimuth/modules/base_classes/dask_module.py]: Good question, I had a conversation about it with @Dref360 but didn’t ask for details. Do we need to add something in the code to allow the GPU to be assigned to worker 0?

File: azimuth/modules/base_classes/dask_module.py (hunk 4 above)
Comment (JosephMarinier) [reply azimuth/modules/base_classes/dask_module.py:40]: Perfect, I think I'm following, then. If https://github.com/ServiceNow/azimuth/pull/525#discussion_r1171396147 works, we can ignore this.

File: azimuth/modules/base_classes/dask_module.py (hunk 8 above)
Comment (gabegma) [reply azimuth/modules/base_classes/dask_module.py:123]: Yeah, I like it!

File: azimuth/modules/base_classes/artifact_manager.py (hunk 7 above)
Comment (JosephMarinier) [reply azimuth/modules/base_classes/artifact_manager.py:163]: > This avoids loading it twice.

Because of eval and train?

Comment (gabegma) [reply azimuth/modules/base_classes/artifact_manager.py:163]: Yes exactly, sorry.

Comment (JosephMarinier) [reply azimuth/modules/base_classes/artifact_manager.py:163]: I get it now. Thanks!

Comment (gabegma) [reply azimuth/modules/base_classes/artifact_manager.py:163]: Yeah, I guess the same question applies to the pipelines and the dataset.

File: azimuth/modules/base_classes/dask_module.py (hunk 5 above)
Comment (Dref360) [reply azimuth/modules/base_classes/dask_module.py]: No it's more of a convention we decided on. All Modules that require GPU **must** be assigned to worker 0.

File: azimuth/modules/base_classes/artifact_manager.py (hunk 7 above)
Comment (Dref360) [reply azimuth/modules/base_classes/artifact_manager.py:163]: Loading the encoder is not a huge cost vs loading the model, but I think this is worthwhile so that in the future we can better control the GPU memory used by Azimuth.

If your model makes 100% of the GPU mem, we are currently not able to load the sentence transformer.

Comment (gabegma) [reply azimuth/modules/base_classes/artifact_manager.py:163]: Do you mean that it is worthwhile to assign tasks that can load the encoder to a different worker than the one that loads the pipeline, or that it is worthwhile to load it once in the artifact manager? Or both?

Comment (gabegma) [reply azimuth/modules/base_classes/artifact_manager.py:163]: Oh really! But without CPU, this overloads one worker no?

Also, I think I would revert my change of moving the encoder to the ArtifactManager, here is the difference in the memory used.
7  assign workers
10  encoder not in AM

File: azimuth/modules/base_classes/dask_module.py [hunk 9]
```diff
@@ -118,6 +125,7 @@ def start_task_on_dataset_split(
             pure=False,
             dependencies=deps,
             key=f"{self.task_id}_{uuid.uuid4()}",  # Unique identifier
+            workers=self.worker,
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/dask_module.py:128]: Do we need to add that to the other `client.submit()` too, in `DaskModule.start_task()`?

Comment (gabegma) [reply azimuth/modules/base_classes/dask_module.py:128]: Good point!!

File: tests/test_modules/test_dataset_analysis/test_class_overlap.py [hunk 10]
```diff
@@ -20,10 +20,10 @@ def encode(self, x, *args, **kwargs):
 
 
 def test_incomplete_class_set(clinc_text_config, dask_client, monkeypatch):
-    monkeypatch.setattr(faiss_mod, "SentenceTransformer", MockedTransformer)
+    monkeypatch.setattr(artifact_manager, "SentenceTransformer", MockedTransformer)
     # Mock SentenceTransformer on all workers.
     dask_client.run(
-        lambda: monkeypatch.setattr(faiss_mod, "SentenceTransformer", MockedTransformer)
+        lambda: monkeypatch.setattr(artifact_manager, "SentenceTransformer", MockedTransformer)
```
Comment (JosephMarinier) [inline tests/test_modules/test_dataset_analysis/test_class_overlap.py:26]: Should we revert the changes in this test file too?

Comment (gabegma) [reply tests/test_modules/test_dataset_analysis/test_class_overlap.py:26]: Yeah already done

Comment (gabegma) [reply tests/test_modules/test_dataset_analysis/test_class_overlap.py:26]: Ah oops no! Fixing :)

**Comment by gabegma** [comment]
BTW - While investigating memory usage - I’m under the impression that we have some memory leakage issues with the saliency tasks. I added issue #549.

**Review by JosephMarinier — APPROVED** [review]
Awesome! Good job on the graphs and all the profiling. 👍

File: azimuth/modules/perturbation_testing/perturbation_testing.py [hunk 11]
```diff
@@ -58,6 +59,9 @@ class PerturbationTestingModule(DatasetResultModule[PerturbationTestingConfig]):
     """
 
     required_mod_options = {"pipeline_index"}
+    # This module doesn't call self.get_model() but requires the model (predict_task.compute(batch))
```
[the remark this answers is not in the document]
Comment (gabegma) [reply azimuth/modules/perturbation_testing/perturbation_testing.py:62]: I caught it just in time :D
